lib/web/desktop_playback.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/websocket"
+)
+
+const actionPlayPause = "play/pause"
+
+const (
+	playStatePlaying  = "playing"
+	playStatePaused   = "paused"
+	playStateFinished = "finished"
+)
+
+// playbackAction is a message passed from the playback client
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionPlayPause | TODO: actionMove
+	// actionPlayPause toggles the playbackState.playState
+	Action string `json:"action"`
+}
+
+// playbackPlayer is a thread-safe struct for managing
+// the global state of the playback websocket connection.
+type playbackPlayer struct {
+	ws        *websocket.Conn
+	playState string // one of playStatePlaying | playStatePaused | playStateFinished
+	cond      *sync.Cond
+	mu        *sync.RWMutex
```

Comment:
```suggestion
	mu        sync.RWMutex
```

Generally you don't want a pointer to mutex.

Reply:
The reason this is a pointer is because that mutex is allocated (via a `var mu sync.RWMutex`) first in `NewPlaybackPlayer` before being passed into the `playbackPlayer` struct. If I don't make it a pointer, I get a `go vet` warning on [line 79](https://github.com/gravitational/teleport/blob/isaiah/session-recording-websocket/lib/web/desktop_playback.go#L79) saying `literal copies lock value from mu: sync.RWMutex (copylocks)`